File: NEsper.Tests/regression/client/TestConfigurationOperations.cs
// ---------------------------------------------------------------------------------- /
// Copyright (C) 2007 Esper Team. All rights reserved.                                /
// http://esper.codehaus.org                                                          /
// ---------------------------------------------------------------------------------- /
// The software in this package is published under the terms of the GPL license       /
// a copy of which has been included with this distribution in the license.txt file.  /
// ---------------------------------------------------------------------------------- /

using System;
using System.Collections.Generic;
using System.IO;
using System.Xml;

using NUnit.Framework;

using net.esper.client;
using net.esper.compat;
using net.esper.support.bean;
using net.esper.support.util;

namespace net.esper.regression.client
{
	[TestFixture]
	public class TestConfigurationOperations
	{
	    private EPServiceProvider epService;
	    private SupportUpdateListener testListener;
	    private ConfigurationOperations configOps;

	    [SetUp]
	    public void SetUp()
	    {
	        testListener = new SupportUpdateListener();
	        epService = EPServiceProviderManager.GetDefaultProvider();
	        epService.Initialize();
	        configOps = epService.EPAdministrator.Configuration;
	    }

	    [Test]
	    public void testAddDOMType()
	    {
	        TryInvalid("AddedDOMOne");

	        // First statement with new name
	        ConfigurationEventTypeXMLDOM domConfig = new ConfigurationEventTypeXMLDOM();
	        domConfig.RootElementName = "RootAddedDOMOne";
	        configOps.AddEventTypeAlias("AddedDOMOne", domConfig);

	        EPStatement stmt = epService.EPAdministrator.CreateEQL("select * from AddedDOMOne");
	        stmt.AddListener(testListener);

	        XmlDocument eventOne = MakeDOMEvent("RootAddedDOMOne");
	        epService.EPRuntime.SendEvent(eventOne);
	        Assert.AreSame(eventOne, testListener.AssertOneGetNewAndReset().Underlying);

	        TryInvalid("AddedMapNameSecond");

	        // Second statement using a new alias to the same type, should both receive
	        domConfig = new ConfigurationEventTypeXMLDOM();
	        domConfig.RootElementName = "RootAddedDOMOne";
	        configOps.AddEventTypeAlias("AddedDOMSecond", domConfig);

	        configOps.AddEventTypeAlias("AddedMapNameSecond", domConfig);
	        SupportUpdateListener testListenerTwo = new SupportUpdateListener();
	        stmt = epService.EPAdministrator.CreateEQL("select * from AddedMapNameSecond");
            stmt.AddListener(testListenerTwo);

            XmlDocument eventTwo = MakeDOMEvent("RootAddedDOMOne");
	        epService.EPRuntime.SendEvent(eventTwo);
	        Assert.IsFalse(testListener.IsInvoked);
	        Assert.AreEqual(eventTwo, testListenerTwo.AssertOneGetNewAndReset().Underlying);

	        // Add the same alias and type again
	        domConfig = new ConfigurationEventTypeXMLDOM();
	        domConfig.RootElementName = "RootAddedDOMOne";
	        configOps.AddEventTypeAlias("AddedDOMSecond", domConfig);

	        // Add the same alias and a different type
	        try
	        {
	            domConfig = new ConfigurationEventTypeXMLDOM();
	            domConfig.RootElementName = "RootAddedDOMXXX";
	            configOps.AddEventTypeAlias("AddedDOMSecond", domConfig);
	            Assert.Fail();
	        }
	        catch (ConfigurationException ex)
	        {
	            // expected
	        }
	    }

	    [Test]
	    public void testAddMapByClass()
	    {
	        TryInvalid("AddedMapOne");

	        // First statement with new name
	        EDictionary<String, Type> mapProps = new HashDictionary<String, Type>();
	        mapProps.Put("prop1", typeof(int));
	        configOps.AddEventTypeAlias("AddedMapOne", mapProps);

	        EPStatement stmt = epService.EPAdministrator.CreateEQL("select * from AddedMapOne");
	        stmt.AddListener(testListener);

            DataDictionary eventOne = new DataDictionary();
	        eventOne.Put("prop1", 1);
	        epService.EPRuntime.SendEvent(eventOne, "AddedMapOne");
	        Assert.AreEqual(eventOne, testListener.AssertOneGetNewAndReset().Underlying);

	        TryInvalid("AddedMapNameSecond");

	        // Second statement using a new alias to the same type, should only one receive
	        configOps.AddEventTypeAlias("AddedMapNameSecond", mapProps);
	        SupportUpdateListener testListenerTwo = new SupportUpdateListener();
	        stmt = epService.EPAdministrator.CreateEQL("select * from AddedMapNameSecond");
            stmt.AddListener(testListenerTwo);

            DataDictionary eventTwo = new DataDictionary();
	        eventTwo.Put("prop1", 1);
	        epService.EPRuntime.SendEvent(eventTwo, "AddedMapNameSecond");
	        Assert.IsFalse(testListener.IsInvoked);
	        Assert.AreEqual(eventTwo, testListenerTwo.AssertOneGetNewAndReset().Underlying);

	        // Add the same alias and type again
	        mapProps.Clear();
	        mapProps.Put("prop1", typeof(int));
	        configOps.AddEventTypeAlias("AddedNameSecond", mapProps);

	        // Add the same alias and a different type
	        try
	        {
	            mapProps.Put("XX", typeof(int));
	            configOps.AddEventTypeAlias("AddedNameSecond", mapProps);
	            Assert.Fail();
	        }
	        catch (ConfigurationException ex)
	        {
	            // expected
	        }
	    }

	    [Test]
	    public void testAddMapProperties()
	    {
	        TryInvalid("AddedMapOne");

	        // First statement with new name
	        Properties mapProps = new Properties();
	        mapProps.Put("prop1", typeof(int).FullName);
	        configOps.AddEventTypeAlias("AddedMapOne", mapProps);

	        EPStatement stmt = epService.EPAdministrator.CreateEQL("select * from AddedMapOne");
	        stmt.AddListener(testListener);

            DataDictionary eventOne = new DataDictionary();
	        eventOne.Put("prop1", 1);
	        epService.EPRuntime.SendEvent(eventOne, "AddedMapOne");
	        Assert.AreEqual(eventOne, testListener.AssertOneGetNewAndReset().Underlying);

	        TryInvalid("AddedMapNameSecond");

	        // Second statement using a new alias to the same type, should only one receive
	        configOps.AddEventTypeAlias("AddedMapNameSecond", mapProps);
	        SupportUpdateListener testListenerTwo = new SupportUpdateListener();
	        stmt = epService.EPAdministrator.CreateEQL("select * from AddedMapNameSecond");
            stmt.AddListener(testListenerTwo);

            DataDictionary eventTwo = new DataDictionary();
	        eventTwo.Put("prop1", 1);
	        epService.EPRuntime.SendEvent(eventTwo, "AddedMapNameSecond");
	        Assert.IsFalse(testListener.IsInvoked);
	        Assert.AreEqual(eventTwo, testListenerTwo.AssertOneGetNewAndReset().Underlying);

	        // Add the same alias and type again
	        mapProps.Clear();
	        mapProps.Put("prop1", typeof(int).FullName);
	        configOps.AddEventTypeAlias("AddedNameSecond", mapProps);

	        // Add the same alias and a different type
	        try
	        {
	            mapProps.Put("XX", typeof(int).FullName);
	            configOps.AddEventTypeAlias("AddedNameSecond", mapProps);
	            Assert.Fail();
	        }
	        catch (ConfigurationException ex)
	        {
	            // expected
	        }
	    }

	    [Test]
	    public void testAddAliasClassName()
	    {
	        TryInvalid("AddedName");

	        // First statement with new name
	        configOps.AddEventTypeAlias("AddedName", typeof(SupportBean).FullName);
	        EPStatement stmt = epService.EPAdministrator.CreateEQL("select * from AddedName");
	        stmt.AddListener(testListener);

	        SupportBean eventOne = new SupportBean("a", 1);
	        epService.EPRuntime.SendEvent(eventOne);
	        Assert.AreSame(eventOne, testListener.AssertOneGetNewAndReset().Underlying);

	        TryInvalid("AddedNameSecond");

	        // Second statement using a new alias to the same type, should both receive
	        configOps.AddEventTypeAlias("AddedNameSecond", typeof(SupportBean).FullName);
	        SupportUpdateListener testListenerTwo = new SupportUpdateListener();
	        stmt = epService.EPAdministrator.CreateEQL("select * from AddedNameSecond");
            stmt.AddListener(testListenerTwo);

	        SupportBean eventTwo = new SupportBean("b", 2);
	        epService.EPRuntime.SendEvent(eventTwo);
	        Assert.AreSame(eventTwo, testListener.AssertOneGetNewAndReset().Underlying);
	        Assert.AreSame(eventTwo, testListenerTwo.AssertOneGetNewAndReset().Underlying);

	        // Add the same alias and type again
	        configOps.AddEventTypeAlias("AddedNameSecond", typeof(SupportBean).FullName);

	        // Add the same alias and a different type
	        try
	        {
	            configOps.AddEventTypeAlias("AddedNameSecond", typeof(SupportBean_A).FullName);
	            Assert.Fail();
	        }
	        catch (ConfigurationException ex)
	        {
	            // expected
	        }
	    }

	    [Test]
	    public void testAddAliasClass()
	    {
	        TryInvalid("AddedName");

	        // First statement with new name
	        configOps.AddEventTypeAlias("AddedName", typeof(SupportBean));
	        EPStatement stmt = epService.EPAdministrator.CreateEQL("select * from AddedName");
	        stmt.AddListener(testListener);

	        SupportBean eventOne = new SupportBean("a", 1);
	        epService.EPRuntime.SendEvent(eventOne);
	        Assert.AreSame(eventOne, testListener.AssertOneGetNewAndReset().Underlying);

	        TryInvalid("AddedNameSecond");

	        // Second statement using a new alias to the same type, should both receive
	        configOps.AddEventTypeAlias("AddedNameSecond", typeof(SupportBean));
	        SupportUpdateListener testListenerTwo = new SupportUpdateListener();
	        stmt = epService.EPAdministrator.CreateEQL("select * from AddedNameSecond");
            stmt.AddListener(testListenerTwo);

	        SupportBean eventTwo = new SupportBean("b", 2);
	        epService.EPRuntime.SendEvent(eventTwo);
	        Assert.AreSame(eventTwo, testListener.AssertOneGetNewAndReset().Underlying);
	        Assert.AreSame(eventTwo, testListenerTwo.AssertOneGetNewAndReset().Underlying);

	        // Add the same alias and type again
	        configOps.AddEventTypeAlias("AddedNameSecond", typeof(SupportBean));

	        // Add the same alias and a different type
	        try
	        {
	            configOps.AddEventTypeAlias("AddedNameSecond", typeof(SupportBean_A));
	            Assert.Fail();
	        }
	        catch (ConfigurationException ex)
	        {
	            // expected
	        }
	    }

	    private void TryInvalid(String alias)
	    {
	        try
	        {
	            epService.EPAdministrator.CreateEQL("select * from " + alias);
	            Assert.Fail();
	        }
	        catch (EPStatementException ex)
	        {
	            // expected
	        }
	    }

	    private XmlDocument MakeDOMEvent(String rootElementName)
	    {
	        String XML =
	            "<VAL1>\n" +
	            "  <someelement/>\n" +
	            "</VAL1>";

	        String xml = XML.Replace("VAL1", rootElementName);

	        XmlDocument document = new XmlDocument() ;
	        document.LoadXml( xml ) ;
	        return document;
	    }
	}
} // End of namespace

File: trunk/esper/src/test/java/net/esper/regression/client/TestConfigurationOperations.java
package net.esper.regression.client;

import net.esper.client.*;
import net.esper.support.util.SupportUpdateListener;
import net.esper.support.bean.SupportBean;
import net.esper.support.bean.SupportBean_A;
import net.esper.support.client.SupportConfigFactory;
import junit.framework.TestCase;

import java.util.Properties;
import java.util.Map;
import java.util.HashMap;
import java.io.StringReader;

import org.xml.sax.InputSource;
import org.w3c.dom.Document;

import javax.xml.parsers.DocumentBuilderFactory;

public class TestConfigurationOperations extends TestCase
{
    private EPServiceProvider epService;
    private SupportUpdateListener testListener;
    private ConfigurationOperations configOps;

    public void setUp()
    {
        testListener = new SupportUpdateListener();
        epService = EPServiceProviderManager.getDefaultProvider(SupportConfigFactory.getConfiguration());
        epService.initialize();
        configOps = epService.getEPAdministrator().getConfiguration();
    }

    public void testAutoAliasPackage()
    {
        configOps.addEventTypeAutoAlias(this.getClass().getPackage().getName());

        EPStatement stmt = epService.getEPAdministrator().createEQL("select * from " + MyAutoAliasEventType.class.getSimpleName());
        stmt.addListener(testListener);

        MyAutoAliasEventType eventOne = new MyAutoAliasEventType(10);
        epService.getEPRuntime().sendEvent(eventOne);
        assertSame(eventOne, testListener.assertOneGetNewAndReset().getUnderlying());
    }

    public void testAutoAliasPackageAmbigous()
    {
        Configuration config = SupportConfigFactory.getConfiguration();
        config.addEventTypeAutoAlias(this.getClass().getPackage().getName());
        configOps.addEventTypeAutoAlias(this.getClass().getPackage().getName());
        configOps.addEventTypeAutoAlias(SupportBean.class.getPackage().getName());

        try
        {
            epService.getEPAdministrator().createEQL("select * from " + SupportAmbigousEventType.class.getSimpleName());
            fail();
        }
        catch (Exception ex)
        {
            assertEquals("Failed to resolve event type: Failed to resolve alias 'SupportAmbigousEventType', the class was ambigously found both in package 'net.esper.regression.client' and in package 'net.esper.support.bean' [select * from SupportAmbigousEventType]", ex.getMessage());
        }

        try
        {
            epService.getEPAdministrator().createEQL("select * from XXXX");
            fail();
        }
        catch (Exception ex)
        {
            assertEquals("Failed to resolve event type: Failed to load class XXXX [select * from XXXX]", ex.getMessage());
        }
    }

    public void testAddDOMType() throws Exception
    {
        tryInvalid("AddedDOMOne");

        // First statement with new name
        ConfigurationEventTypeXMLDOM domConfig = new ConfigurationEventTypeXMLDOM();
        domConfig.setRootElementName("RootAddedDOMOne");
        configOps.addEventTypeAlias("AddedDOMOne", domConfig);

        EPStatement stmt = epService.getEPAdministrator().createEQL("select * from AddedDOMOne");
        stmt.addListener(testListener);

        Document eventOne = makeDOMEvent("RootAddedDOMOne");
        epService.getEPRuntime().sendEvent(eventOne);
        assertSame(eventOne, testListener.assertOneGetNewAndReset().getUnderlying());

        tryInvalid("AddedMapNameSecond");

        // Second statement using a new alias to the same type, should both receive
        domConfig = new ConfigurationEventTypeXMLDOM();
        domConfig.setRootElementName("RootAddedDOMOne");
        configOps.addEventTypeAlias("AddedDOMSecond", domConfig);

        configOps.addEventTypeAlias("AddedMapNameSecond", domConfig);
        SupportUpdateListener testListenerTwo = new SupportUpdateListener();
        stmt = epService.getEPAdministrator().createEQL("select * from AddedMapNameSecond");
        stmt.addListener(testListenerTwo);

        Document eventTwo = makeDOMEvent("RootAddedDOMOne");
        epService.getEPRuntime().sendEvent(eventTwo);
        assertTrue(testListener.isInvoked());
        assertEquals(eventTwo, testListenerTwo.assertOneGetNewAndReset().getUnderlying());

        // Add the same alias and type again
        domConfig = new ConfigurationEventTypeXMLDOM();
        domConfig.setRootElementName("RootAddedDOMOne");
        configOps.addEventTypeAlias("AddedDOMSecond", domConfig);

        // Add the same alias and a different type
        try
        {
            domConfig = new ConfigurationEventTypeXMLDOM();
            domConfig.setRootElementName("RootAddedDOMXXX");
            configOps.addEventTypeAlias("AddedDOMSecond", domConfig);
            fail();
        }
        catch (ConfigurationException ex)
        {
            // expected
        }
    }

    public void testAddMapByClass()
    {
        tryInvalid("AddedMapOne");

        // First statement with new name
        Map<String, Class> mapProps = new HashMap<String, Class>();
        mapProps.put("prop1", int.class);
        configOps.addEventTypeAlias("AddedMapOne", mapProps);

        EPStatement stmt = epService.getEPAdministrator().createEQL("select * from AddedMapOne");
        stmt.addListener(testListener);

        Map<String, Object> eventOne = new HashMap<String, Object>();
        eventOne.put("prop1", 1);
        epService.getEPRuntime().sendEvent(eventOne, "AddedMapOne");
        assertEquals(eventOne, testListener.assertOneGetNewAndReset().getUnderlying());

        tryInvalid("AddedMapNameSecond");

        // Second statement using a new alias to the same type, should only one receive
        configOps.addEventTypeAlias("AddedMapNameSecond", mapProps);
        SupportUpdateListener testListenerTwo = new SupportUpdateListener();
        stmt = epService.getEPAdministrator().createEQL("select * from AddedMapNameSecond");
        stmt.addListener(testListenerTwo);

        Map<String, Object> eventTwo = new HashMap<String, Object>();
        eventTwo.put("prop1", 1);
        epService.getEPRuntime().sendEvent(eventTwo, "AddedMapNameSecond");
        assertFalse(testListener.isInvoked());
        assertEquals(eventTwo, testListenerTwo.assertOneGetNewAndReset().getUnderlying());

        // Add the same alias and type again
        mapProps.clear();
        mapProps.put("prop1", int.class);
        configOps.addEventTypeAlias("AddedNameSecond", mapProps);

        // Add the same alias and a different type
        try
        {
            mapProps.put("XX", int.class);
            configOps.addEventTypeAlias("AddedNameSecond", mapProps);
            fail();
        }
        catch (ConfigurationException ex)
        {
            // expected
        }
    }

    public void testAddMapProperties()
    {
        tryInvalid("AddedMapOne");

        // First statement with new name
        Properties mapProps = new Properties();
        mapProps.put("prop1", int.class.getName());
        configOps.addEventTypeAlias("AddedMapOne", mapProps);

        EPStatement stmt = epService.getEPAdministrator().createEQL("select * from AddedMapOne");
        stmt.addListener(testListener);

        Map<String, Object> eventOne = new HashMap<String, Object>();
        eventOne.put("prop1", 1);
        epService.getEPRuntime().sendEvent(eventOne, "AddedMapOne");
        assertEquals(eventOne, testListener.assertOneGetNewAndReset().getUnderlying());

        tryInvalid("AddedMapNameSecond");

        // Second statement using a new alias to the same type, should only one receive
        configOps.addEventTypeAlias("AddedMapNameSecond", mapProps);
        SupportUpdateListener testListenerTwo = new SupportUpdateListener();
        stmt = epService.getEPAdministrator().createEQL("select * from AddedMapNameSecond");
        stmt.addListener(testListenerTwo);

        Map<String, Object> eventTwo = new HashMap<String, Object>();
        eventTwo.put("prop1", 1);
        epService.getEPRuntime().sendEvent(eventTwo, "AddedMapNameSecond");
        assertFalse(testListener.isInvoked());
        assertEquals(eventTwo, testListenerTwo.assertOneGetNewAndReset().getUnderlying());

        // Add the same alias and type again
        mapProps.clear();
        mapProps.put("prop1", int.class.getName());
        configOps.addEventTypeAlias("AddedNameSecond", mapProps);

        // Add the same alias and a different type
        try
        {
            mapProps.put("XX", int.class.getName());
            configOps.addEventTypeAlias("AddedNameSecond", mapProps);
            fail();
        }
        catch (ConfigurationException ex)
        {
            // expected
        }
    }

    public void testAddAliasClassName()
    {
        tryInvalid("AddedName");

        // First statement with new name
        configOps.addEventTypeAlias("AddedName", SupportBean.class.getName());
        EPStatement stmt = epService.getEPAdministrator().createEQL("select * from AddedName");
        stmt.addListener(testListener);

        SupportBean eventOne = new SupportBean("a", 1);
        epService.getEPRuntime().sendEvent(eventOne);
        assertSame(eventOne, testListener.assertOneGetNewAndReset().getUnderlying());

        tryInvalid("AddedNameSecond");

        // Second statement using a new alias to the same type, should both receive
        configOps.addEventTypeAlias("AddedNameSecond", SupportBean.class.getName());
        SupportUpdateListener testListenerTwo = new SupportUpdateListener();
        stmt = epService.getEPAdministrator().createEQL("select * from AddedNameSecond");
        stmt.addListener(testListenerTwo);

        SupportBean eventTwo = new SupportBean("b", 2);
        epService.getEPRuntime().sendEvent(eventTwo);
        assertSame(eventTwo, testListener.assertOneGetNewAndReset().getUnderlying());
        assertSame(eventTwo, testListenerTwo.assertOneGetNewAndReset().getUnderlying());
        
        // Add the same alias and type again
        configOps.addEventTypeAlias("AddedNameSecond", SupportBean.class.getName());

        // Add the same alias and a different type
        try
        {
            configOps.addEventTypeAlias("AddedNameSecond", SupportBean_A.class.getName());
            fail();
        }
        catch (ConfigurationException ex)
        {
            // expected
        }
    }

    public void testAddAliasClass()
    {
        tryInvalid("AddedName");

        // First statement with new name
        configOps.addEventTypeAlias("AddedName", SupportBean.class);
        EPStatement stmt = epService.getEPAdministrator().createEQL("select * from AddedName");
        stmt.addListener(testListener);

        SupportBean eventOne = new SupportBean("a", 1);
        epService.getEPRuntime().sendEvent(eventOne);
        assertSame(eventOne, testListener.assertOneGetNewAndReset().getUnderlying());

        tryInvalid("AddedNameSecond");

        // Second statement using a new alias to the same type, should both receive
        configOps.addEventTypeAlias("AddedNameSecond", SupportBean.class);
        SupportUpdateListener testListenerTwo = new SupportUpdateListener();
        stmt = epService.getEPAdministrator().createEQL("select * from AddedNameSecond");
        stmt.addListener(testListenerTwo);

        SupportBean eventTwo = new SupportBean("b", 2);
        epService.getEPRuntime().sendEvent(eventTwo);
        assertSame(eventTwo, testListener.assertOneGetNewAndReset().getUnderlying());
        assertSame(eventTwo, testListenerTwo.assertOneGetNewAndReset().getUnderlying());

        // Add the same alias and type again
        configOps.addEventTypeAlias("AddedNameSecond", SupportBean.class);

        // Add the same alias and a different type
        try
        {
            configOps.addEventTypeAlias("AddedNameSecond", SupportBean_A.class);
            fail();
        }
        catch (ConfigurationException ex)
        {
            // expected
        }
    }

    private void tryInvalid(String alias)
    {
        try
        {
            epService.getEPAdministrator().createEQL("select * from " + alias);
            fail();
        }
        catch (EPStatementException ex)
        {
            // expected
        }
    }

    private Document makeDOMEvent(String rootElementName) throws Exception
    {
        String XML =
            "<VAL1>\n" +
            "  <someelement/>\n" +
            "</VAL1>";

        String xml = XML.replaceAll("VAL1", rootElementName);

        StringReader reader = new StringReader(xml);
        InputSource source = new InputSource(reader);
        DocumentBuilderFactory builderFactory = DocumentBuilderFactory.newInstance();
        builderFactory.setNamespaceAware(true);
        Document simpleDoc = builderFactory.newDocumentBuilder().parse(source);
        return simpleDoc;
    }
}
